pkg/assembler/backends/inmem/hashEqual.go

```diff
@@ -169,47 +156,72 @@ func (c *demoClient) matchArtifacts(filter []*model.ArtifactSpec, value []uint32
 
 // Query HashEqual
 
-func (c *demoClient) HashEqual(ctx context.Context, hSpec *model.HashEqualSpec) ([]*model.HashEqual, error) {
-	if len(hSpec.Artifacts) > 2 {
+func (c *demoClient) HashEqual(ctx context.Context, filter *model.HashEqualSpec) ([]*model.HashEqual, error) {
+	funcName := "HashEqual"
+	if len(filter.Artifacts) > 2 {
 		return nil, gqlerror.Errorf(
-			"HashEqual :: Provided spec has too many Artifacts")
+			"%v :: Provided spec has too many Artifacts", funcName)
 	}
 
-	// If ID is provided, try to look up, then check if rest matches
-	if hSpec.ID != nil {
-		id64, err := strconv.ParseUint(*hSpec.ID, 10, 32)
+	if filter.ID != nil {
+		id64, err := strconv.ParseUint(*filter.ID, 10, 32)
 		if err != nil {
-			return nil, gqlerror.Errorf("HashEqual :: invalid ID %s", err)
+			return nil, gqlerror.Errorf("%v :: invalid ID %s", funcName, err)
 		}
 		id := uint32(id64)
-		h, err := c.hashEqualByID(id)
+		link, err := byID[*hashEqualStruct](id, c)
 		if err != nil {
 			// Not found
 			return nil, nil
 		}
 		// If found by id, ignore rest of fields in spec and return as a match
-		return []*model.HashEqual{c.convHashEqual(h)}, nil
+		return []*model.HashEqual{c.convHashEqual(link)}, nil
 	}
 
-	var hashEquals []*model.HashEqual
-	// TODO if any artifacts are exact matches only search those backedges
-	for _, h := range c.hashEquals {
-		if noMatch(hSpec.Justification, h.justification) ||
-			noMatch(hSpec.Origin, h.origin) ||
-			noMatch(hSpec.Collector, h.collector) ||
-			!c.matchArtifacts(hSpec.Artifacts, h.artifacts) {
-			continue
+	var search []uint32
+	foundOne := false
+	for _, a := range filter.Artifacts {
+		if !foundOne && a != nil {
+			exactArtifact, err := c.artifactExact(a)
+			if err != nil {
+				return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+			}
+			if exactArtifact != nil {
+				search = append(search, exactArtifact.hashEquals...)
+				foundOne = true
+				break
+			}
 		}
-		hashEquals = append(hashEquals, c.convHashEqual(h))
 	}
 
-	return hashEquals, nil
+	var out []*model.HashEqual
+	if foundOne {
+		for _, id := range search {
+			link, err := byID[*hashEqualStruct](id, c)
+			if err != nil {
+				return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+			}
+			out, err = c.addHEIfMatch(out, filter, link)
+			if err != nil {
+				return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+			}
+		}
+	} else {
```

Comment:
hmm if one side was not found, should this return not found?

Reply:
Yes, if there is an exact match on a subject/object then we don't need a fallback to search all links, only search the backedges and if none match, it will fall through and return an empty list.